src/supertux/sector.cpp

```diff
@@ -1003,68 +1004,74 @@ Sector::handle_collisions()
+      && moving_object->get_group() != COLGROUP_MOVING_ONLY_STATIC)
+     || !moving_object->is_valid())
+    continue;
 
-    uint32_t tile_attributes = collision_tile_attributes(moving_object->dest, moving_object->get_movement());
-    if(tile_attributes >= Tile::FIRST_INTERESTING_FLAG) {
-      moving_object->collision_tile(tile_attributes);
-    }
+  uint32_t tile_attributes = collision_tile_attributes(moving_object->dest, moving_object->get_movement());
+  if(tile_attributes >= Tile::FIRST_INTERESTING_FLAG) {
+    moving_object->collision_tile(tile_attributes);
   }
-
-  // part2.5: COLGROUP_MOVING vs COLGROUP_TOUCHABLE
-  for(const auto& moving_object : moving_objects) {
-    if((moving_object->get_group() != COLGROUP_MOVING
-        && moving_object->get_group() != COLGROUP_MOVING_STATIC)
-       || !moving_object->is_valid())
+}
+for(const auto& moving_object : moving_objects)
+  c.insert(moving_object);
+
+// part2.5: COLGROUP_MOVING vs COLGROUP_TOUCHABLE
+for(const auto& moving_object : moving_objects) {
+  if((moving_object->get_group() != COLGROUP_MOVING
+      && moving_object->get_group() != COLGROUP_MOVING_STATIC)
+     || !moving_object->is_valid())
+    continue;
+  c.retrieve(moving_object,fill);
+  for(auto& moving_object_2 : fill) {
+    if(moving_object_2->get_group() != COLGROUP_TOUCHABLE
+       || !moving_object_2->is_valid())
       continue;
 
-    for(auto& moving_object_2 : moving_objects) {
-      if(moving_object_2->get_group() != COLGROUP_TOUCHABLE
-         || !moving_object_2->is_valid())
+    if(intersects(moving_object->dest, moving_object_2->dest)) {
+      Vector normal;
+      CollisionHit hit;
+      get_hit_normal(moving_object->dest, moving_object_2->dest,
+                     hit, normal);
+      if(!moving_object->collides(*moving_object_2, hit))
+        continue;
+      if(!moving_object_2->collides(*moving_object, hit))
         continue;
 
-      if(intersects(moving_object->dest, moving_object_2->dest)) {
-        Vector normal;
-        CollisionHit hit;
-        get_hit_normal(moving_object->dest, moving_object_2->dest,
-                       hit, normal);
-        if(!moving_object->collides(*moving_object_2, hit))
-          continue;
-        if(!moving_object_2->collides(*moving_object, hit))
-          continue;
-
-        moving_object->collision(*moving_object_2, hit);
-        moving_object_2->collision(*moving_object, hit);
-      }
+      moving_object->collision(*moving_object_2, hit);
+      moving_object_2->collision(*moving_object, hit);
     }
   }
+}
 
-  // part3: COLGROUP_MOVING vs COLGROUP_MOVING
-  for(auto i = moving_objects.begin(); i != moving_objects.end(); ++i) {
-    auto moving_object = *i;
-
-    if((moving_object->get_group() != COLGROUP_MOVING
-        && moving_object->get_group() != COLGROUP_MOVING_STATIC)
-       || !moving_object->is_valid())
+// part3: COLGROUP_MOVING vs COLGROUP_MOVING
+for(auto i = moving_objects.begin(); i != moving_objects.end(); ++i) {
```

Comment:
Now that you decoupled the i2 iterator from the i iterator, both can use the shorter syntax :)